File: leetcode/1001-1999/1456-maximum-number-of-vowels-in-a-substring-of-given-length/MaximumNumberOfVowelsInASubstringOfGivenLength.ts
function maxVowels(s: string, k: number): number {
  const evaluate = c => "aoeiu".indexOf(c) < 0 ? 0 : 1;

  const length = s.length;

  let max = 0;

  for (let i = 0; i < k; i++) {
    max += evaluate(s[i]);
  }

  let current = max;
  console.log(`current=${current}, max=${max}`);

  for (let i = k; i < length && max < k; i++) {
    console.log(`current=${current}, max=${max}`);
    current = current + evaluate(s[i]) - evaluate(s[i - k]);
    max = Math.max(current, max);
    console.log(`current=${current}, max=${max}`);
  }


  return max;
};

console.log(maxVowels("leetcode", 3))

File: leetcode/1001-2000/1401-1500/1456-maximum-number-of-vowels-in-a-substring-of-given-length/MaximumNumberOfVowelsInASubstringOfGivenLength.java
public class MaximumNumberOfVowelsInASubstringOfGivenLength {
  public static void main(String[] args) {
    System.out.println(new Solution().maxVowels("leetcode", 3));
  }
}

class Solution {
  public int maxVowels(String s, int k) {
    if (s.isEmpty() || k == 0) {
      return 0;
    }

    int max = 0;
    final String vowels = "aoeiu";

    for (int i = 0; i < k; i++) {
      if (vowels.indexOf(s.charAt(i)) < 0) {
        continue;
      }

      max++;
    }

    int current = max;

    for (int i = k; i < s.length() && max < k; i++) {
      if (vowels.indexOf(s.charAt(i - k)) >= 0) {
        current--;
      }

      if (vowels.indexOf(s.charAt(i)) < 0) {
        continue;
      }

      current++;

      max = Math.max(current, max);
    }

    return max;
  }
}
